examples/Iris/step3_optimize.cc

```diff
@@ -0,0 +1,413 @@
+                                                                     {"PW", {&IrisFlower::PW, 3u, "Petal.Width"}}};
+static std::map<std::string, Feature> features(features_list.begin(), features_list.end());
+
+static std::map<std::string, size_t> label_indexes{{"setosa", 0u}, {"versicolor", 1u}, {"virginica", 2u}};
+
+struct NoModelVisualizer {
+  template <typename T>
+  NoModelVisualizer(T&&) {}
+  void AddPlots(current::gnuplot::GNUPlot& unused_plot, size_t unused_dimension_x, size_t unused_dimension_y) const {
+    static_cast<void>(unused_plot);
+    static_cast<void>(unused_dimension_x);
+    static_cast<void>(unused_dimension_y);
+  }
+};
+
+template <typename T = NoModelVisualizer>
+Response Plot(const std::vector<IrisFlower>& flowers,
+              const std::string& x,
+              const std::string& y,
+              bool nolegend,
+              size_t image_size,
+              double point_size,
+              T&& model_visualizer = T(),
+              const std::string& output_format = "svg",
+              const std::string& content_type = current::net::constants::kDefaultSVGContentType) {
+  try {
+    const auto ix = features.at(x);
+    const auto iy = features.at(y);
+    const double IrisFlower::*px = ix.mem_ptr;
+    const double IrisFlower::*py = iy.mem_ptr;
+    using namespace current::gnuplot;
+    GNUPlot plot;
+    if (nolegend) {
+      plot.NoBorder().NoTics().NoKey();
+    } else {
+      plot.Title("Iris Data").Grid("back").XLabel(features.at(x).name).YLabel(features.at(y).name);
+    }
+    plot.ImageSize(image_size).OutputFormat(output_format);
+    for (const auto& label : labels) {
+      auto plot_data = WithMeta([&label, &flowers, px, py](Plotter& p) {
+        for (const auto& flower : flowers) {
+          if (flower.Label == label.name) {
+            p(flower.*px, flower.*py);
+          }
+        }
+      });
+      plot.Plot(plot_data.AsPoints()
+                    .Name(label.name)
+                    .Color(label.color)
+                    .Extra("pt " + label.pt + (point_size ? "ps " + current::ToString(point_size) : "")));
+    }
+    model_visualizer.AddPlots(plot, ix.dimension_index, iy.dimension_index);
+    return Response(static_cast<std::string>(plot), HTTPResponseCode.OK, content_type);
+  } catch (const std::out_of_range&) {
+    return Response("Invalid dimension.", HTTPResponseCode.BadRequest);
+  }
+}
+
+enum class ComputationType { TrainDescriptiveModel, TrainDiscriminantModel, ComputeAccuracy };
+
+template <typename T>
+struct WeightsComputer {
+  T inv_r_squared[3];
+  T k[3];
+
+  // For a 4D Gaussian, `exp(-(distance_squared / radius_squared))`, the normalization coefficient
+  // is `radius_squared ^ (-2)`. As an implementation detail, use inverted values.
+  explicit WeightsComputer(const std::vector<T>& x) {
+    for (size_t c = 0; c < 3; ++c) {
+      inv_r_squared[c] = fncas::exp(x[c * 5 + 4]);
```

Comment:
If you store radius as `-log(r)`, this one is just `inv_r`, where is the `squared` part?